rclc/src/functions.c

```diff
@@ -0,0 +1,235 @@
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc = rcl_wait_set_init(&wait_set, 1, 0, 0, 0, 0, rcl_get_default_allocator());
+  if(rc != RCL_RET_OK) {
+    puts("ERROR : rcl_wait_set_init");
+  }
+
+  rc = rcl_wait_set_clear_subscriptions(&wait_set);
+  if(rc != RCL_RET_OK) {
+    puts("ERROR : rcl_wait_set_clear_subscriptions");
+  }
+
+  rc = rcl_wait(&wait_set, RCL_MS_TO_NS(milliseconds));
+  if (rc == RCL_RET_TIMEOUT) {
+    rc = rcl_wait_set_fini(&wait_set);
+    if(rc != RCL_RET_OK) {
+      puts("ERROR : rcl_wait_set_fini");
+    }
+    return;
+  }
+
+  rc = rcl_wait_set_fini(&wait_set);
+  if(rc != RCL_RET_OK) {
+    puts("ERROR : rcl_wait_set_fini");
+  }
+}
+
+void rclc_spin_node(rclc_node_t * node)
+{
+  (void)node;
+
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc = rcl_wait_set_init(&wait_set, 1, 0, 0, 0, 0, rcl_get_default_allocator());
+  if(rc != RCL_RET_OK) {
+    puts("ERROR : rcl_wait_set_init");
+  }
+
+  while(1) {
+    rc = rcl_wait_set_clear_subscriptions(&wait_set);
+    if(rc != RCL_RET_OK) {
+      puts("ERROR : rcl_wait_set_clear_subscriptions");
+    }
+
+    for (size_t i = 0; i < node->subs_s; ++i) {
+      rc = rcl_wait_set_add_subscription(&wait_set, node->subs[i]->rcl_subscription);
+      if(rc != RCL_RET_OK) {
+        puts("ERROR : rcl_wait_set_add_subscription");
+      }
+    }
+
+    rc = rcl_wait(&wait_set, RCL_MS_TO_NS(1000));  // 1000ms == 1s, passed as ns
+    if (rc == RCL_RET_TIMEOUT) {
+      continue;
+    }
+
+    for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+      if (wait_set.subscriptions[i]) {
+        char msg[64];
+        rc = rcl_take(wait_set.subscriptions[i], (void*)&msg, NULL);
+        if(rc != RCL_RET_OK) {
+          puts("ERROR : rcl_take");
+        }
+        else {
+          for (size_t i = 0; i < node->subs_s; ++i) {
+            if(node->subs[i]->rcl_subscription == wait_set.subscriptions[i]) {
+              node->subs[i]->user_callback((const void*)&msg);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  rc = rcl_wait_set_fini(&wait_set);
+  if(rc != RCL_RET_OK) {
+    puts("ERROR : rcl_wait_set_fini");
+  }
+}
+
+rclc_node_t * rclc_create_node(const char * name)
+{
+  const char* namespace_ = ""; // TODO : put as parameter
+
+  rclc_node_t* ret = malloc(sizeof(rclc_node_t));
+  memset(ret, 0, sizeof(rclc_node_t));
+
+  {
+    rcl_node_t node = rcl_get_zero_initialized_node();
+    ret->rcl_node = malloc(sizeof(rcl_node_t));
+    memcpy(ret->rcl_node, &node, sizeof(rcl_node_t));
+  }
+  
+  rcl_node_options_t node_ops = rcl_node_get_default_options();
+  rcl_ret_t rc = rcl_node_init(ret->rcl_node, name, namespace_, &node_ops);
+  if(rc != RCL_RET_OK) {
+    puts("ERROR : rcl_node_init");
+    return NULL;
+  }
+  
+  return ret;
+}
+
+rclc_ret_t rclc_destroy_node(rclc_node_t * node)
+{
+  rcl_ret_t ret = rcl_node_fini(node->rcl_node);
+  free(node);
```

Comment:
Symmetric to the above: `rcl_get_default_allocator()->deallocate` instead of `free`.